### PR TITLE
[cli] Implement `dagster definitions validate`

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR implements `dagster definitions validate`, a command to validate the definitions of a Dagster project.

---

### Expected behavior: 

Have a command to validate the definitions of a Dagster project. A behavior similar to [Definitions.validate_loadable](https://github.com/dagster-io/dagster/blob/e755c47bc8e1f596fdb4c4d68551732ee656dcfb/python_modules/dagster/dagster/_core/definitions/definitions_class.py#L632-L643) is expected.

Example:

```
> dagster definitions validate -f path/to/my/definitions.py

Validating definitions in path/to/my/definitions.py
...

> dagster definitions validate -m my_dagster_project.definitions

Validating definitions in my_dagster_project.definitions.
...

> dagster definitions validate

Validating definitions in ...
...
```

### Input:

Similar to `dagster dev` as described [here](https://docs.dagster.io/guides/running-dagster-locally#locating-your-code) - the definitions can be validated:
- from a file, specified as a command line argument, eg. `definitions.py`
- from a module, specified as a command line argument, eg. `my_dagster_project.definitions`
- without command line arguments, read from `pyproject.toml` or `workspace.yml`.

### Output:

Initially as logs in stdout/stderr.

### In scope:
- Support the validation of Dagster definitions defined in a `Definitions` object.
- Implement a behavior similar to `Definitions.validate_loadable` where the enclosed definitions will be loadable by Dagster:
  - No assets have conflicting keys.
  - No jobs, sensors, or schedules have conflicting names.
  - All asset jobs can be resolved.
  - All resource requirements are satisfied.

### Out of scope:
- Support the validation of Dagster definitions defined in a `Repository` object.
  - See comment [here](https://dagsterlabs.slack.com/archives/C06C9QVJBUL/p1723762251730029?thread_ts=1723761841.463129&cid=C06C9QVJBUL)
- Validate definitions dependencies across code locations
  - See thread [here](https://dagsterlabs.slack.com/archives/C06C9QVJBUL/p1723642633214359)
  - Could be added in a next iteration.

### Potential improvement and next steps:
- Support usage in CI/CD
- Support output as JSON
- Support definitions constraints
  - eg. Sensors' `minimum_interval_seconds` should always be of a at least 120 seconds.
- Support validation of definitions with dependencies across code locations.

## How I Tested These Changes

Tested locally
BK with additional tests

## Changelog

Add the `dagster definitions validate` command to Dagster CLI. This command validates if Dagster definitions are loadable.

- [x] `NEW` _(added new feature or capability)_
- [ ] `BUGFIX` _(fixed a bug)_
- [ ] `DOCS` _(added or updated documentation)_

### CODE REVIEW

**Comment by maximearmstrong** [comment]
* **#23692** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23692?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maximearmstrong and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_cli/check.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            python_file=python_file,
+            package_name=None,
+        )
+
+    # TODO: Raise exception is loadable_target_origin is None?
+
+    logger.info("Starting check...")
+
+    # TODO: Logging when loading the repository would be great, where should it be implemented?
+    LoadedRepositories(
```
Comment (sryza) [inline python_modules/dagster/dagster/_cli/check.py:76]: In our other workspace-level CLI commands, just use `get_workspace_from_kwargs` I believe. Could we use the same here?

**Review by sryza — COMMENTED** [review]
Seems very useful!

Main feedback is on naming. `dagster check` makes me think it would be executing my asset checks. `validate-loadable` is more of a mouthful, but is more consistent with the python operation of the same name, if I understand what this is doing correctly? or `validate-defs`?

**Comment by maximearmstrong** [comment]
> Main feedback is on naming. dagster check makes me think it would be executing my asset checks. validate-loadable is more of a mouthful, but is more consistent with the python operation of the same name, if I understand what this is doing correctly? or validate-defs?

@sryza Definitely - we've reached a similar conclusion in [this thread](https://dagsterlabs.slack.com/archives/C04PD3G2F7H/p1724165051724099?thread_ts=1724097103.960319&cid=C04PD3G2F7H).

Our current consensus is `dagster validate`. If we want to be more specific, I think we could go with something like `dagster validate definitions`, but are we considering validating anything other than definitions? Under the hood we validate definitions, but the intention here is to validate that a Dagster project is valid and loadable.

File: python_modules/dagster/dagster/_cli/check.py (hunk 1 above)
Comment (maximearmstrong) [reply python_modules/dagster/dagster/_cli/check.py:76]: I believe so. This initial draft should be simplified and adapted to match the other workspace-level CLI commands.

**Comment by sryza** [comment]
> Our current consensus is dagster validate. If we want to be more specific, I think we could go with something like dagster validate definitions, but are we considering validating anything other than definitions? Under the hood we validate definitions, but the intention here is to validate that a Dagster project is valid and loadable.

I think it's worth starting a thread in #eng-naming, but I suspect we should probably be more specific. You could imagine other functionality that e.g. validates attributes of a running deployment.

File: python_modules/dagster/dagster/_cli/definitions.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                loadable_target_origin=code_location_origin.loadable_target_origin,
+                entry_point=DEFAULT_DAGSTER_ENTRY_POINT,
+            )
+        except Exception as e:
+            click.echo(
+                click.style("Validation failed with exception: ", fg="red") + f"{e}.", err=True
+            )
+            sys.exit(1)
+        else:
+            click.echo("Validation successful!")
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_cli/definitions.py:66]: One thing that I don't like about this is that it loads all the different code locations in the same process. It's not a given that if you have two different code locations that they will play nicely together - they might have inconsistent packages that don't work together for example.

Would it be possible to load each of these up in a subprocess like the webserver does on startup? And pass along the load error over gRPC for any of them if there are any?

**Review by gibsondan — CHANGES_REQUESTED** [review]
one high level piece of feedback inline - can we load these up in subprocesses like the webserver and daemon do? There should already be machinery in place to do that - you can call get_workspace_from_kwargs like a lot of our existing CLI commands do, and it will do all the work of trying to load up each of the code locations for you. Then you can iterate through each of them and raise an exception if any of them are in an error state?

File: python_modules/dagster/dagster/_cli/definitions.py (hunk 2 above)
Comment (maximearmstrong) [reply python_modules/dagster/dagster/_cli/definitions.py]: That makes perfect sense - done in [655b3a6](https://github.com/dagster-io/dagster/pull/23692/commits/655b3a651eaa62f36d2784cbdb21a3fb06d0f748)

**Review by sryza — APPROVED** [review]

**Pull request merged**